cpp/src/datetime/datetime_ops.cu

```diff
@@ -88,32 +90,60 @@ struct extract_component_operator {
   }
 };
 
-template <datetime_component COMPONENT>
-struct ceil_timestamp {
+// This functor takes the rounding type as runtime info and dispatches to the ceil/floor/round
+// function.
+template <typename DurationType>
+struct RoundFunctor {
+  template <typename Timestamp>
+  CUDA_DEVICE_CALLABLE auto operator()(rounding_kind round_kind, Timestamp dt)
+  {
+    switch (round_kind) {
+      case rounding_kind::CEIL: return cuda::std::chrono::ceil<DurationType>(dt);
+      case rounding_kind::FLOOR: return cuda::std::chrono::floor<DurationType>(dt);
+      case rounding_kind::ROUND: return cuda::std::chrono::round<DurationType>(dt);
+      default: cudf_assert(false && "Unsupported rounding kind.");
```

Comment:
@vyasr does this mean the `__builtin_unreachable();` after the default case can be removed now, essentially only having this once?

Reply:
No, although I think we should probably make a decision project-wide on how to handle default cases if we're running a switch conditioned on an enum. Without a default clause the code will just not execute anything. If this is still useful for debugging purposes maybe we just leave it as is. I'm going to resolve this comment for now since on second thought I don't think it's something we want to deal with in this PR.

Reply:
Is `__builtin_unreachable()` always appropriate for the `default:` of a switch that exhausts its enum's cases? I was under the impression that this was only meant to be used at the _end_ of the function to eliminate errors about not returning, and that the default case is still good to keep. Having a failure here would prevent problems if an enum is updated but some of the places that enum is used are not updated. I think I'd rather crash the program with an error message in that situation than have undefined behavior.

Reply:
This is discussed in #7753. I'm going to define a macro unreachable code in a follow-up PR as suggested in that issue. @vyasr Should we leave this as it is for now and then I'll unify the behavior later? The current approach of `default: cudf_assert` followed by `__builtin_unreachable()` after the `switch` seems fine for the interim?

Reply:
Yeah skip it for now. This change will be better done in concert with all unreachable-related changes.